### DISCUSSION TITLE
Try to train vocoder, checkpoint['optimizer']

### DISCUSSION THREAD

**Discussion by Raizenr** [coqui-ai/TTS#490] [General Q&A]
I try to start train pre-trained hifigan vocoder model, but i have some problems with checkpoint['optimizer'].
I downloaded https://github.com/coqui-ai/TTS/releases "English | VCTK | HiFiGAN | vocoder | 😄v0.0.13" and used my tatar language data. 

python /app/data/TTS/TTS/bin/distribute.py --script train_vocoder_gan.py --config_path /app/data/hifigan-transformers-edresson/config.json --restore_path /app/data/hifigan-transformers-edresson/model_file.pth.tar

How can i fix it? I tried to use stats path, it doesn't work. I am newbie. 

config.json:

        {
        "run_name": "hifigan",
        "run_description": "hifigan mean-var scaling",

    // AUDIO PARAMETERS
    "audio":{
        // Audio processing parameters
        "num_mels": 80,         // size of the mel spec frame.
        "fft_size": 1024,       // number of stft frequency levels. Size of the linear spectogram frame.
        "sample_rate": 22050, //22050,   // DATASET-RELATED: wav sample-rate. If different than the original data, it is resampled.
        "win_length": 1024,     // stft window length in ms.
        "hop_length": 256,      // stft window hop-lengh in ms.
        "frame_length_ms": null,  // stft window length in ms.If null, 'win_length' is used.
        "frame_shift_ms": null,   // stft window hop-lengh in ms. If null, 'hop_length' is used.
        "preemphasis": 0.0,    // pre-emphasis to reduce spec noise and make it more structured. If 0.0, no -pre-emphasis.
        "min_level_db": -100,   // normalization range
        "ref_level_db": 20,     // reference level db, theoretically 20db is the sound of air.
        "power": 1.5,           // value to sharpen wav signals after GL algorithm.
        "griffin_lim_iters": 60,// #griffin-lim iterations. 30-60 is a good range. Larger the value, slower the generation.
        "stft_pad_mode": "reflect",
        // Normalization parameters
        "signal_norm": true,    // normalize the spec values in range [0, 1]
        "symmetric_norm": true, // move normalization to range [-1, 1]
        "max_norm": 4.0,          // scale normalization to range [-max_norm, max_norm] or [0, max_norm]
        "clip_norm": true,      // clip normalized values into the range.
        "mel_fmin": 0.0,         // minimum freq level for mel-spec. ~50 for male and ~95 for female voices. Tune for dataset!!
        "mel_fmax": 8000.0,        // maximum freq level for mel-spec. Tune for dataset!!
        "spec_gain": 1.0,
        "do_trim_silence": false,  // enable trimming of slience of audio as you load it. LJspeech (false), TWEB (false), Nancy (true)
        "trim_db": 60,          // threshold for timming silence. Set this according to your dataset.
        "stats_path": null
    },

    // DISTRIBUTED TRAINING
    "distributed":{
        "backend": "nccl",
        "url": "tcp:\/\/localhost:54324"
    },

    // MODEL PARAMETERS
    "use_pqmf": false,

    // LOSS PARAMETERS
    "use_stft_loss": false,
    "use_subband_stft_loss": false,
    "use_mse_gan_loss": true,
    "use_hinge_gan_loss": false,
    "use_feat_match_loss": false,  // use only with melgan discriminators
    "use_l1_spec_loss": true,

    // loss weights
     "stft_loss_weight": 0,
     "subband_stft_loss_weight": 0,
     "mse_G_loss_weight": 1,
     "hinge_G_loss_weight": 0,
     "feat_match_loss_weight": 10,
     "l1_spec_loss_weight": 45,

    // multiscale stft loss parameters
    // "stft_loss_params": {
    //     "n_ffts": [1024, 2048, 512],
    //     "hop_lengths": [120, 240, 50],
    //     "win_lengths": [600, 1200, 240]
    // },

    "l1_spec_loss_params": {
        "use_mel": true,
        "sample_rate": 22050,
        "n_fft": 1024,
        "hop_length": 256,
        "win_length": 1024,
        "n_mels": 80,
        "mel_fmin": 0.0,
        "mel_fmax": null
    },

    "target_loss": "avg_G_loss",  // loss value to pick the best model to save after each epoch

    // DISCRIMINATOR
     "discriminator_model": "hifigan_discriminator",
     //"discriminator_model_params":{
     //    "peroids": [2, 3, 5, 7, 11],
     //    "base_channels": 16,
     //    "max_channels":512,
     //    "downsample_factors":[4, 4, 4]
     //},
    "steps_to_start_discriminator": 0,      // steps required to start GAN trainining.1
    "diff_samples_for_G_and_D": false,   // draw a new sample from the dataset for the D pass.

    // GENERATOR
    "generator_model": "hifigan_generator",
    "generator_model_params": {
        "upsample_factors":[8,8,2,2],
        "upsample_kernel_sizes": [16,16,4,4],
        "upsample_initial_channel": 128,
        "resblock_kernel_sizes": [3,7,11],
        "resblock_dilation_sizes": [[1,3,5], [1,3,5], [1,3,5]],
        "resblock_type": "1"
    },

    // DATASET
    "data_path": "/app/data/resampled_model_22050",
    "feature_path": null,
    // "feature_path": "/home/erogol/gdrive/Datasets/non-binary-voice-files/tacotron-DCA/",
    "seq_len": 8192,
    "pad_short": 2000,
    "conv_pad": 0,
    "use_noise_augment": false,
    "use_cache": true,
    "reinit_layers": [],    // give a list of layer names to restore from the given checkpoint. If not defined, it reloads all heuristically matching layers.
	 // TRAINING
    "batch_size": 32,       // Batch size for training. Lower values than 32 might cause hard to learn attention. It is overwritten by 'gradual_training'.

    // VALIDATION
    "run_eval": true,
    "test_delay_epochs": 10,  //Until attention is aligned, testing only wastes computation time.
    "test_sentences_file": null,  // set a file to load sentences to be used for testing. If it is null then we use default english sentences.

    // OPTIMIZER
    "epochs": 10000,                // total number of epochs to train.
    "wd": 0.0,                // Weight decay weight.
    "gen_clip_grad": -1,      // Generator gradient clipping threshold. Apply gradient clipping if > 0
    "disc_clip_grad": -1,     // Discriminator gradient clipping threshold.
    "lr_gen": 0.0002,                  // Initial learning rate. If Noam decay is active, maximum learning rate.
    "lr_disc": 0.0002,
    "optimizer": "AdamW",
    "optimizer_params":{
        "betas": [0.8, 0.99],
        "weight_decay": 0.0
    },
    "lr_scheduler_gen": "ExponentialLR",   // one of the schedulers from https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
    "lr_scheduler_gen_params": {
       "gamma": 0.999,
        "last_epoch": -1
    },
    "lr_scheduler_disc": "ExponentialLR",   // one of the schedulers from https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
    "lr_scheduler_disc_params": {
          "gamma": 0.999,
      // TENSORBOARD and LOGGING
     "print_step": 25,       // Number of steps to log traning on console.
     "print_eval": false,     // If True, it prints loss values for each step in eval run.
     "save_step": 25000,      // Number of training steps expected to plot training stats on TB and save model checkpoints.
     "checkpoint": true,     // If true, it saves checkpoints per "save_step"
     "tb_model_param_stats": true,     // true, plots param stats per layer on tensorboard. Might be memory consuming, but good for debugging.

    // DATA LOADING
    "num_loader_workers": 8,        // number of training data loader processes. Don't set it too big. 4-8 are good values.
    "num_val_loader_workers": 4,    // number of evaluation data loader processes.
    "eval_split_size": 10,

    // PATHS
    "output_path": "/app/data/tatar_vocoder_1/"
     }

console:

  root@e82bbb799dd4:/app/data# python /app/data/TTS/TTS/bin/distribute.py --script train_vocoder_gan.py --config_path 
  /app/data/hifigan-transformers-edresson/config.json --restore_path /app/data/hifigan-transformers-edresson/model_file.pth.tar
  ['/app/data/TTS/TTS/bin/train_vocoder_gan.py', '--continue_path=', '--restore_path=/app/data/hifigan-transformers- 
  edresson/model_file.pth.tar', '--config_path=/app/data/hifigan-transformers-edresson/config.json', '--group_id=group_2021_05_19- 
  091441', '--rank=0']
  ['/app/data/TTS/TTS/bin/train_vocoder_gan.py', '--continue_path=', '--restore_path=/app/data/hifigan-transformers- 
  edresson/model_file.pth.tar', '--config_path=/app/data/hifigan-transformers-edresson/config.json', '--group_id=group_2021_05_19- 
  091441', '--rank=1']
  ['/app/data/TTS/TTS/bin/train_vocoder_gan.py', '--continue_path=', '--restore_path=/app/data/hifigan-transformers- 
  edresson/model_file.pth.tar', '--config_path=/app/data/hifigan-transformers-edresson/config.json', '--group_id=group_2021_05_19- 
  091441', '--rank=2']
  ['/app/data/TTS/TTS/bin/train_vocoder_gan.py', '--continue_path=', '--restore_path=/app/data/hifigan-transformers- 
  edresson/model_file.pth.tar', '--config_path=/app/data/hifigan-transformers-edresson/config.json', '--group_id=group_2021_05_19- 
  091441', '--rank=3']
  ['/app/data/TTS/TTS/bin/train_vocoder_gan.py', '--continue_path=', '--restore_path=/app/data/hifigan-transformers- 
  edresson/model_file.pth.tar', '--config_path=/app/data/hifigan-transformers-edresson/config.json', '--group_id=group_2021_05_19- 
  091441', '--rank=4']
  ['/app/data/TTS/TTS/bin/train_vocoder_gan.py', '--continue_path=', '--restore_path=/app/data/hifigan-transformers- 
  edresson/model_file.pth.tar', '--config_path=/app/data/hifigan-transformers-edresson/config.json', '--group_id=group_2021_05_19- 
  091441', '--rank=5']
  ['/app/data/TTS/TTS/bin/train_vocoder_gan.py', '--continue_path=', '--restore_path=/app/data/hifigan-transformers- 
  edresson/model_file.pth.tar', '--config_path=/app/data/hifigan-transformers-edresson/config.json', '--group_id=group_2021_05_19- 
  091441', '--rank=6']
  ['/app/data/TTS/TTS/bin/train_vocoder_gan.py', '--continue_path=', '--restore_path=/app/data/hifigan-transformers- 
  edresson/model_file.pth.tar', '--config_path=/app/data/hifigan-transformers-edresson/config.json', '--group_id=group_2021_05_19- 
  091441', '--rank=7']
fatal: not a git repository (or any parent up to mount point /app)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
fatal: not a git repository (or any parent up to mount point /app)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
fatal: not a git repository (or any parent up to mount point /app)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
fatal: not a git repository (or any parent up to mount point /app)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
 > Using CUDA:  True
 > Number of GPUs:  8
fatal: not a git repository (or any parent up to mount point /app)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
 > Git Hash: 0000000
 > Experiment folder: /app/data/tatar_vocoder_1/hifigan-May-19-2021_09+14AM-0000000
fatal: not a git repository (or any parent up to mount point /app)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
fatal: not a git repository (or any parent up to mount point /app)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
 > Loading wavs from: /app/data/resampled_model_22050
fatal: not a git repository (or any parent up to mount point /app)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
fatal: not a git repository (or any parent up to mount point /app)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
 > Setting up Audio Processor...
 | > sample_rate:22050
 | > resample:False
 | > num_mels:80
 | > min_level_db:-100
 | > frame_shift_ms:None
 | > frame_length_ms:None
 | > ref_level_db:20
 | > fft_size:1024
 | > power:1.5
 | > preemphasis:0.0
 | > griffin_lim_iters:60
 | > signal_norm:True
 | > symmetric_norm:True
 | > mel_fmin:0
 | > mel_fmax:8000.0
 | > spec_gain:1.0
 | > stft_pad_mode:reflect
 | > max_norm:4.0
 | > clip_norm:True
 | > do_trim_silence:False
 | > trim_db:60
 | > do_sound_norm:False
 | > stats_path:None
 | > base:10
 | > hop_length:256
 | > win_length:1024
 > Generator Model: hifigan_generator
 > Discriminator Model: hifigan_discriminator
Traceback (most recent call last):
  File "/app/data/TTS/TTS/bin/train_vocoder_gan.py", line 622, in <module>
    main(args)
  File "/app/data/TTS/TTS/bin/train_vocoder_gan.py", line 517, in main
    optimizer_gen.load_state_dict(checkpoint["optimizer"])
  File "/usr/local/lib/python3.6/dist-packages/torch/optim/optimizer.py", line 116, in load_state_dict
    saved_groups = state_dict['param_groups']
TypeError: 'NoneType' object is not subscriptable
Traceback (most recent call last):
  File "/app/data/TTS/TTS/bin/train_vocoder_gan.py", line 622, in <module>
    main(args)
  File "/app/data/TTS/TTS/bin/train_vocoder_gan.py", line 517, in main
    optimizer_gen.load_state_dict(checkpoint["optimizer"])
  File "/usr/local/lib/python3.6/dist-packages/torch/optim/optimizer.py", line 116, in load_state_dict
    saved_groups = state_dict['param_groups']
TypeError: 'NoneType' object is not subscriptable
 > Restoring from model_file.pth.tar...
Traceback (most recent call last):
  File "/app/data/TTS/TTS/bin/train_vocoder_gan.py", line 622, in <module>
    main(args)
  File "/app/data/TTS/TTS/bin/train_vocoder_gan.py", line 517, in main
    optimizer_gen.load_state_dict(checkpoint["optimizer"])
  File "/usr/local/lib/python3.6/dist-packages/torch/optim/optimizer.py", line 116, in load_state_dict
    saved_groups = state_dict['param_groups']
TypeError: 'NoneType' object is not subscriptable
 > Restoring Generator Model...
Traceback (most recent call last):
  File "/app/data/TTS/TTS/bin/train_vocoder_gan.py", line 622, in <module>
    main(args)
  File "/app/data/TTS/TTS/bin/train_vocoder_gan.py", line 517, in main
    optimizer_gen.load_state_dict(checkpoint["optimizer"])
  File "/usr/local/lib/python3.6/dist-packages/torch/optim/optimizer.py", line 116, in load_state_dict
    saved_groups = state_dict['param_groups']
TypeError: 'NoneType' object is not subscriptable
Traceback (most recent call last):
  File "/app/data/TTS/TTS/bin/train_vocoder_gan.py", line 622, in <module>
    main(args)
  File "/app/data/TTS/TTS/bin/train_vocoder_gan.py", line 517, in main
    optimizer_gen.load_state_dict(checkpoint["optimizer"])
  File "/usr/local/lib/python3.6/dist-packages/torch/optim/optimizer.py", line 116, in load_state_dict
    saved_groups = state_dict['param_groups']
TypeError: 'NoneType' object is not subscriptable
 > Restoring Generator Optimizer...
Traceback (most recent call last):
  File "/app/data/TTS/TTS/bin/train_vocoder_gan.py", line 622, in <module>
    main(args)
  File "/app/data/TTS/TTS/bin/train_vocoder_gan.py", line 517, in main
    optimizer_gen.load_state_dict(checkpoint["optimizer"])
  File "/usr/local/lib/python3.6/dist-packages/torch/optim/optimizer.py", line 116, in load_state_dict
    saved_groups = state_dict['param_groups']
TypeError: 'NoneType' object is not subscriptable
Traceback (most recent call last):
  File "/app/data/TTS/TTS/bin/train_vocoder_gan.py", line 622, in <module>
    main(args)
  File "/app/data/TTS/TTS/bin/train_vocoder_gan.py", line 517, in main
    optimizer_gen.load_state_dict(checkpoint["optimizer"])
  File "/usr/local/lib/python3.6/dist-packages/torch/optim/optimizer.py", line 116, in load_state_dict
    saved_groups = state_dict['param_groups']
TypeError: 'NoneType' object is not subscriptable
Traceback (most recent call last):
  File "/app/data/TTS/TTS/bin/train_vocoder_gan.py", line 622, in <module>
    main(args)
  File "/app/data/TTS/TTS/bin/train_vocoder_gan.py", line 517, in main
    optimizer_gen.load_state_dict(checkpoint["optimizer"])
  File "/usr/local/lib/python3.6/dist-packages/torch/optim/optimizer.py", line 116, in load_state_dict
    saved_groups = state_dict['param_groups']
TypeError: 'NoneType' object is not subscriptable

**Comment by erogol** [comment]
I guess you dont have git installed in your system? Try installing `git` and follow the installation steps described on the project README.md

**Reply by Raizenr** [reply]
I have already installed git with all steps from README.md
Before that, I trained the Tacotron 2 model

**Comment by lexkoro** [comment]
I think the model you are trying to fine-tune on is only for inference (since it's only 3MB in size). 
So it doesn't have the state of the optimizer saved, which is why you get the error i guess?

**Reply by erogol** [reply]
good point. It is true. It has only the generator.

**Reply by Raizenr** [reply]
Probably, I was advised to train voсoder using an already trained English language model. 
I tried to off optimizer but it failed to start train too
